backend/src/main/java/ai/verta/modeldb/lineage/LineageDAORdbImpl.java

```diff
@@ -30,36 +49,241 @@
-          deleteLineage(session, input, output);
+        deleteLineage(session, input, inputInDatabase, blobHashInCommitFunction);
+      }
+      for (LineageEntry output : deleteLineage.getOutputList()) {
+        deleteLineage(session, output, outputInDatabase, blobHashInCommitFunction);
+      }
+      status = false;
+      boolean noDataProvided =
+          deleteLineage.getInputCount() == 0 && deleteLineage.getOutputCount() == 0;
+      if (noDataProvided || inputInDatabase.isEmpty() || outputInDatabase.isEmpty()) {
+        Set<ConnectionEntity> values = new HashSet<>(inputInDatabase.values());
+        values.addAll(outputInDatabase.values());
+        values.forEach(connectionEntity -> deleteConnectionEntity(session, connectionEntity));
+        LineageElementEntity lineageElementEntity = session.get(LineageElementEntity.class, id);
+        if (lineageElementEntity != null) {
+          session.remove(lineageElementEntity);
+          status = true;
         }
       }
       session.getTransaction().commit();
     }
-    return DeleteLineage.Response.newBuilder().setStatus(true).build();
+    return DeleteLineage.Response.newBuilder().setStatus(status).build();
+  }
+
+  private Entry<Map<LineageElement, ConnectionEntity>, Map<LineageElement, ConnectionEntity>>
+      getInputOutputs(Session session, List result) throws ModelDBException {
+    Map<LineageElement, ConnectionEntity> inputInDatabase = new HashMap<>();
+    Map<LineageElement, ConnectionEntity> outputInDatabase = new HashMap<>();
+    for (Object entity : result) {
+      ConnectionEntity connectionEntity = (ConnectionEntity) entity;
+      if (connectionEntity.getConnectionType() == CONNECTION_TYPE_INPUT) {
+        inputInDatabase.put(connectionEntity.getLineageElement(session), connectionEntity);
+      } else {
+        outputInDatabase.put(connectionEntity.getLineageElement(session), connectionEntity);
+      }
+    }
+    return new AbstractMap.SimpleEntry<>(inputInDatabase, outputInDatabase);
+  }
+
+  private Map<LineageElement, ConnectionEntity> getInputOrOutput(Session session, List result)
+      throws ModelDBException {
+    Map<LineageElement, ConnectionEntity> inputOrOutputInDatabase = new HashMap<>();
+    for (Object entity : result) {
+      ConnectionEntity connectionEntity = (ConnectionEntity) entity;
+      inputOrOutputInDatabase.put(connectionEntity.getLineageElement(session), connectionEntity);
+    }
+    return inputOrOutputInDatabase;
+  }
+
+  private List getConnectionEntitiesById(Session session, long id) {
+    return getConnectionEntitiesByIdAndConnectionType(session, id, CONNECTION_TYPE_ANY);
+  }
+
+  private List getConnectionEntitiesByIdAndConnectionType(
+      Session session, long id, int connectionType) {
+    String queryString = "from " + ConnectionEntity.class.getSimpleName() + " where id = " + id;
+    if (connectionType != CONNECTION_TYPE_ANY) {
+      queryString += " and connectionType = " + connectionType;
+    }
```

Comment:
Are we actually passing values **inside** the query instead of in arguments? This seems like an issue waiting to happen. @ravishetye we should scan our code for more issues like this.

Reply:
fixed